cluster/channel.go

```diff
@@ -0,0 +1,130 @@
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// Channel allows clients to send messages for a specific state type that will be
+// broadcasted in a best-effort manner.
+type Channel struct {
+	key          string
+	send         func([]byte)
+	peers        func() []*memberlist.Node
+	sendOversize func(*memberlist.Node, []byte) error
+
+	msgc   chan []byte
+	logger log.Logger
+
+	oversizeGossipMessageFailureTotal prometheus.Counter
+	oversizeGossipDuration            prometheus.Histogram
+}
+
+// NewChannel creates a new Channel struct, which handles sending normal and
+// oversize messages to peers.
+func NewChannel(
+	key string,
+	send func([]byte),
+	peers func() []*memberlist.Node,
+	sendOversize func(*memberlist.Node, []byte) error,
+	logger log.Logger,
+	stopc chan struct{},
+	reg prometheus.Registerer,
+) *Channel {
+	oversizeGossipMessageFailureTotal := prometheus.NewCounter(prometheus.CounterOpts{
+		Name:        "alertmanager_oversized_gossip_message_failure_total",
+		Help:        "Number notification log received queries that failed.",
+		ConstLabels: prometheus.Labels{"key": key},
+	})
+	oversizeGossipDuration := prometheus.NewHistogram(prometheus.HistogramOpts{
+		Name:        "alertmanager_oversize_gossip_message_duration_seconds",
+		Help:        "Duration of oversized gossip message requests.",
+		ConstLabels: prometheus.Labels{"key": key},
+	})
+
+	reg.MustRegister(oversizeGossipDuration, oversizeGossipMessageFailureTotal)
+
+	c := &Channel{
+		key:                               key,
+		send:                              send,
+		peers:                             peers,
+		logger:                            logger,
+		msgc:                              make(chan []byte, 200),
+		sendOversize:                      sendOversize,
+		oversizeGossipMessageFailureTotal: oversizeGossipMessageFailureTotal,
+		oversizeGossipDuration:            oversizeGossipDuration,
+	}
+
+	go c.handleOverSizedMessages(stopc)
+
+	return c
+}
+
+// handleOverSizedMessages prevents memberlist from opening too many parallel
+// TCP connections to its peers.
+func (c *Channel) handleOverSizedMessages(stopc chan struct{}) {
+	var wg sync.WaitGroup
+	for {
+		select {
+		case b := <-c.msgc:
+			for _, n := range c.peers() {
+				wg.Add(1)
+				go func(n *memberlist.Node) {
+					defer wg.Done()
+
+					start := time.Now()
+					if err := c.sendOversize(n, b); err != nil {
+						level.Debug(c.logger).Log("msg", "failed to send reliable", "key", c.key, "node", n, "err", err)
+						c.oversizeGossipMessageFailureTotal.Inc()
+						return
+					}
+					c.oversizeGossipDuration.Observe(time.Since(start).Seconds())
+				}(n)
+			}
+
+			wg.Wait()
+		case <-stopc:
+			return
+		}
+	}
+}
+
+// Broadcast enqueues a message for broadcasting.
+func (c *Channel) Broadcast(b []byte) {
+	b, err := proto.Marshal(&clusterpb.Part{Key: c.key, Data: b})
+	if err != nil {
+		return
+	}
+
+	if OversizedMessage(b) {
+		c.msgc <- b
```

Comment:
This could block if the channel is full, isn't it? In which case, we need this:

```go
select {
    case c.msgc <- b:
    default:
        level.Debug(c.logger).Log("msg", "channel full")
}
```